### PR TITLE
JP-3553: Update background step handling of multi-ints exposures

### LINKED ISSUE #8313: level2b for mir_4qpm jw03254-c1009 crashed with "could not broadcast input array"
_Issue [JP-3553](https://jira.stsci.edu/browse/JP-3553) was created on JIRA by [Hien Tran](https://jira.stsci.edu/secure/ViewProfile.jspa?name=htran):_

ops saw all 9 exposures of the mir_4qpm dataset jw03254-c1009_20240118t185403_image2_0000* (for observation 4) crash in level2b processing with b10.0, in the bkg_subtract step, 
```java
2024-02-26 16:50:47,709 - CRDS - DEBUG - Final effective context is 'jwst_1202.pmap'
2024-02-26 16:50:47,709 - CRDS - DEBUG - Computing best references locally.
2024-02-26 16:50:47,710 - CRDS - DEBUG - Bestrefs header:
{'META.EXPOSURE.NGROUPS [NGROUPS]': '100.0',
'META.EXPOSURE.READPATT [READPATT]': 'FASTR1',
'META.EXPOSURE.TYPE [EXP_TYPE]': 'MIR_4QPM',
'META.INSTRUMENT.BAND [BAND]': 'UNDEFINED',
'META.INSTRUMENT.CHANNEL [CHANNEL]': 'UNDEFINED',
'META.INSTRUMENT.CORONAGRAPH [CORONMSK]': '4QPM_1140',
'META.INSTRUMENT.DETECTOR [DETECTOR]': 'MIRIMAGE',
'META.INSTRUMENT.FILTER [FILTER]': 'F1140C',
'META.INSTRUMENT.LAMP_STATE [LAMP]': 'OFF',
'META.INSTRUMENT.NAME [INSTRUME]': 'MIRI',
'META.OBSERVATION.DATE [DATE-OBS]': '2024-02-26',
'META.OBSERVATION.TIME [TIME-OBS]': '03:11:49.739',
'META.SUBARRAY.NAME [SUBARRAY]': 'MASK1140',
'META.VISIT.TSOVISIT [TSOVISIT]': 'F',
'META.VISIT.TYPE [VISITYPE]': 'PRIME_TARGETED_FIXED',
'REFTYPE': 'UNDEFINED'}
2024-02-26 16:50:47,712 - CRDS - DEBUG - Reference type 'distortion' defined as 'jwst_miri_distortion_0047.asdf'
2024-02-26 16:50:47,712 - CRDS - DEBUG - Reference type 'drizpars' defined as 'jwst_miri_drizpars_0001.fits'
2024-02-26 16:50:47,712 - CRDS - DEBUG - Reference type 'filteroffset' defined as 'jwst_miri_filteroffset_0010.asdf'
2024-02-26 16:50:47,712 - CRDS - DEBUG - Reference type 'flat' defined as 'jwst_miri_flat_0771.fits'
2024-02-26 16:50:47,712 - CRDS - DEBUG - Reference type 'photom' defined as 'jwst_miri_photom_0181.fits'
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for AREA reference file is 'N/A'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for CAMERA reference file is 'N/A'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for COLLIMATOR reference file is 'N/A'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for DFLAT reference file is 'N/A'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for DISPERSER reference file is 'N/A'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for DISTORTION reference file is '/ifs/archive/ops/jwst/ref/tmp_crds/crds/cache/references/jwst/miri/jwst_miri_distortion_0047.asdf'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for DRIZPARS reference file is '/ifs/archive/ops/jwst/ref/tmp_crds/crds/cache/references/jwst/miri/jwst_miri_drizpars_0001.fits'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for FFLAT reference file is 'N/A'.
2024-02-26 16:50:47,714 - stpipe.Image2Pipeline - INFO - Prefetch for FILTEROFFSET reference file is '/ifs/archive/ops/jwst/ref/tmp_crds/crds/cache/references/jwst/miri/jwst_miri_filteroffset_0010.asdf'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for FLAT reference file is '/ifs/archive/ops/jwst/ref/tmp_crds/crds/cache/references/jwst/miri/jwst_miri_flat_0771.fits'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for FORE reference file is 'N/A'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for FPA reference file is 'N/A'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for IFUFORE reference file is 'N/A'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for IFUPOST reference file is 'N/A'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for IFUSLICER reference file is 'N/A'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for MSA reference file is 'N/A'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for OTE reference file is 'N/A'.
2024-02-26 16:50:47,715 - stpipe.Image2Pipeline - INFO - Prefetch for PHOTOM reference file is '/ifs/archive/ops/jwst/ref/tmp_crds/crds/cache/references/jwst/miri/jwst_miri_photom_0181.fits'.
2024-02-26 16:50:47,716 - stpipe.Image2Pipeline - INFO - Prefetch for REGIONS reference file is 'N/A'.
2024-02-26 16:50:47,716 - stpipe.Image2Pipeline - INFO - Prefetch for SFLAT reference file is 'N/A'.
2024-02-26 16:50:47,716 - stpipe.Image2Pipeline - INFO - Prefetch for SPECWCS reference file is 'N/A'.
2024-02-26 16:50:47,716 - stpipe.Image2Pipeline - INFO - Prefetch for WAVELENGTHRANGE reference file is 'N/A'.
2024-02-26 16:50:47,716 - stpipe.Image2Pipeline - INFO - Prefetch for WFSSBKG reference file is 'N/A'.
2024-02-26 16:50:47,717 - stpipe.Image2Pipeline - INFO - Starting calwebb_image2 ...
2024-02-26 16:50:47,724 - stpipe.Image2Pipeline - INFO - Processing product jw03254004001_04101_00006_mirimage
2024-02-26 16:50:47,724 - stpipe.Image2Pipeline - INFO - Working on input jw03254004001_04101_00006_mirimage_rateints.fits ...
2024-02-26 16:50:47,892 - stpipe.Image2Pipeline.bkg_subtract - INFO - Step bkg_subtract running with args (<CubeModel(6, 224, 288) from jw03254004001_04101_00006_mirimage_rateints.fits>, ['jw03254002001_02101_00001_mirimage_rateints.fits', 'jw03254002001_02101_00002_mirimage_rateints.fits', 'jw03254003001_02101_00001_mirimage_rateints.fits', 'jw03254003001_02101_00002_mirimage_rateints.fits']).
2024-02-26 16:50:47,893 - stpipe.Image2Pipeline.bkg_subtract - INFO - Step bkg_subtract parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': 'bsubints', 'search_output_file': True, 'input_dir': '/ifs/archive/ops/jwst/info/owlmgr/paths/sdp/asn_creation/cal/level2', 'save_combined_background': False, 'sigma': 3.0, 'maxiters': None, 'wfss_mmag_extract': None}
2024-02-26 16:50:47,924 - stpipe.Image2Pipeline.bkg_subtract - INFO - Accumulate bkg from jw03254002001_02101_00001_mirimage_rateints.fits
Traceback (most recent call last):
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/bin/strun", line 26, in <module>
step = Step.from_cmdline(sys.argv[1:])
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/stpipe/step.py", line 186, in from_cmdline
return cmdline.step_from_cmdline(args)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/stpipe/cmdline.py", line 386, in step_from_cmdline
step.run(*positional)
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/stpipe/step.py", line 478, in run
step_result = self.process(*args)
^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/jwst/pipeline/calwebb_image2.py", line 67, in process
result = self.process_exposure_product(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/jwst/pipeline/calwebb_image2.py", line 148, in process_exposure_product
input = self.bkg_subtract(input, members_by_type['background'])
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/stpipe/step.py", line 478, in run
step_result = self.process(*args)
^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/jwst/background/background_step.py", line 91, in process
bkg_model, result = background_sub.background_sub(input_model,
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/jwst/background/background_sub.py", line 153, in background_sub
bkg_model = average_background(input_model,
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/jwst/background/background_sub.py", line 222, in average_background
bkg_data, bkg_err, bkg_dq = im_array.get_subset_array(bkg_array)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/jwst/background/background_sub.py", line 112, in get_subset_array
data_overlap[:data_cutout.shape[0], :data_cutout.shape[1]] = copy.deepcopy(data_cutout)
~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: could not broadcast input array from shape (26,224,288) into shape (6,224,288)
2024057165048 INFO src=ssb_calibration_wrapper._strun_analyze_and_log_failure fsn=jw03254-c1009_20240118t185403_image2_00004_asn msg="strun /dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.12.5.20231019-py3.11/lib/python3.11/site-packages/jwst/pipeline/calwebb_image2.cfg FAILED (exit=1) on jw03254-c1009_20240118t185403_image2_00004_asn.json."
2024057165048 ERROR src=ssb_calibration_wrapper.calibrate._strun._strun_analyze_and_log_failure fsn=jw03254-c1009_20240118t185403_image2_00004_asn msg="strun exit status=1" ```
the error message _"could not broadcast input array from shape (26,224,288) into shape (6,224,288)"_ suggests that the numbers of integrations for the background and science images do not match.

the association files for one of the c1009 exposures is attached, along with one for c1008, which did **not** crash. note:
 * c1009 refers to association for obs 4, which is the psf ref star, with nints = 6
 * c1008 is the association for obs 1, which is the actual science target, with nints = 26
 * they both include **two** sets of observations as background members, obs 2 & 3
 * obs 2 has nints = 26, obs 3 has nints = 6
 * c1009 image2 crashed but c1008 did not

questions:
 # are the "background" associations such as these supposed to have more than one set of background observation? it *appears* bkgr obs 2 is intended to go along with obs 1, and bkgr obs 3 with obs 4
 # could the failure be due to the **ordering** of the background members in the asn table (json) file? as obs 2 bkgr member is always listed first after the science member, it matches nints (26) with that of obs 1, but not obs 4

### PR DESCRIPTION
**Pull request by hbushouse**
Resolves [JP-3553](https://jira.stsci.edu/browse/JP-3553)

Closes #8313 

This PR updates the background subtraction step to make it more flexible in handling multi-integration background exposures (rateints files). Previously, the way the code was implemented, it was necessary to have NINTS be the same for the science exposure and all of the background exposures. This update allows the background exposures to have any value of NINTS and still be used correctly to form a mean background, which is subsequently subtracted from every integration of the science exposure. The problem with the previous code was that it was allocating internal arrays for working with the background images based on the shape of the science data, so if they different NINTS, the shape is wrong. This update simply allocates the internal background arrays based on the shape of each background image.

I've also incorporated the data that originally triggered this work into a new form of the existing regression test that runs the background step on multi-integration exposures. All of the necessary input and truth files have been uploaded to Artifactory and the updated test runs successfully on my local branch.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by hbushouse** [comment]
Regression test started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1252

**Comment by hbushouse** [comment]
> Regression test started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1252

Regtest run had 2 unrelated failures. So this PR is clean.

File: CHANGES.rst [hunk 1]
```diff
@@ -10,6 +10,12 @@ associations
 
 - Isolate candidate processing into their own pools [#8227]
 
+background
+----------
+
+- Updated to allow multi-integration (rateints) background exposures to have
+  different values of NINTS. [#8326]
```
Comment (emolter) [inline CHANGES.rst:17]:
```suggestion
  a different value of NINTS than the science exposure. [#8326]
```

**Review by emolter — APPROVED** [review]
Looks good, just suggesting to slightly modify changelog wording

**Pull request merged**